### PR TITLE
Added new methods to SubjectAreaRelationshipRESTResource for CategoryHierarchyLink.

### PR DESCRIPTION
**Pull request by immmus**
Hi! I noticed that SubjectAreaRelationshipRESTResource do not have rest methods for CategoryHierarchyLink.
And I decided that I needed to add them. I wrote rest methods, added to client them and tests.
I hope that the pr to be to merge faster than usually :).

### CODE REVIEW

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 1]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
+     * <li> PropertyServerException              Property server exception. </li>
+     * </ul>
+     */
+    public SubjectAreaOMASAPIResponse<CategoryHierarchyLink> updateCategoryHierarchyLink(String serverName, String userId, String guid, CategoryHierarchyLink categoryHierarchyLink, Boolean isReplace) {
+        String restAPIName = "updateCategoryHierarchyLink";
+        return updateLine(serverName,restAPIName, userId, guid, CategoryHierarchyLinkMapper.class, categoryHierarchyLink, isReplace);
+    }
+
+    /**
+     * Create a CategoryHierarchyLink Relationship. A relationship between a Category and a Category.
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1963]: It is more than a relationship between a category and another category, it is the relationship between a child category and its parent category. I think it is defined in the UML as "Relationship between two categories used to create nested categories."

 It is possible to create this relationship by specifying a category with a parentCategory

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java [hunk 2]
```diff
@@ -198,6 +199,9 @@ public void run() throws SubjectAreaFVTCheckedException, InvalidParameterExcepti
         projectFVT.deleteProject(project.getSystemAttributes().getGUID());
         projectFVT.purgeProject(project.getSystemAttributes().getGUID());
 
+        Category cat3 = catFVT.createCategory(DEFAULT_TEST_CAT_NAME3, glossaryGuid);
+        Category cat4 = catFVT.createCategory(DEFAULT_TEST_CAT_NAME4, glossaryGuid);
+        categoryHierarchyLinkFVT(cat3, cat4);
     }
```
Comment (davidradl) [inline open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:205]: Could we get cat3 and validate that it has a parent of cat4 as well please.

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java [hunk 3]
```diff
@@ -1423,4 +1427,36 @@ private ProjectScope createProjectScope(Project project, Term term) throws Inval
...
+
+    private void categoryHierarchyLinkFVT(Category parent, Category child) throws UserNotAuthorizedException, PropertyServerException, InvalidParameterException, SubjectAreaFVTCheckedException {
+        CategoryHierarchyLink categoryHierarchyLink = createCategoryHierarchyLink(parent, child);
+        String guid = categoryHierarchyLink.getGuid();
+        CategoryHierarchyLink gotCategoryHierarchyLink = subjectAreaRelationship.categoryHierarchyLink().getByGUID(this.userId, guid);
+        FVTUtils.validateLine(gotCategoryHierarchyLink);
+        System.out.println("Got CategoryHierarchyLink " + categoryHierarchyLink);
+        subjectAreaRelationship.categoryHierarchyLink().delete(this.userId, guid);
+        //FVTUtils.validateLine(gotTermCategorizationRelationship);
```
Comment (davidradl) [inline open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:1438]: remove

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 4]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+                                                                                      @PathVariable String guid) {
+        return restAPI.getCategoryHierarchyLink(serverName, userId, guid);
+    }
+
+    /**
+     * Update a CategoryHierarchyLink Relationship. A relationship between a Category and a Category.
+     *
+     * @param serverName           serverName under which this request is performed, this is used in multi tenanting to identify the tenant
+     * @param userId               userId under which the request is performed
+     * @param guid       guid of the CategoryAnchor relationship
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:1990]: should be CategoryHierarchyLink. formatting does not look right.

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 5]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+     * @param categoryHierarchyLink the categoryAnchor relationship
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:1991]: should be CategoryHierarchyLink

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 6]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+     * @return response, when successful contains the updated categoryHierarchyLink
+     * when not successful the following Exception responses can occur
+     * <ul>
+     * <li> UserNotAuthorizedException           the requesting user is not authorized to issue this request.</li>
+     * <li> InvalidParameterException            one of the parameters is null or invalid.</li>
+     * <li> PropertyServerException              Property server exception. </li>
+     * </ul>
+     */
+    @PutMapping(path = "/users/{userId}/relationships/category-hierarchy-link/{guid}")
+    public SubjectAreaOMASAPIResponse<CategoryHierarchyLink> updateCategoryAnchor(@PathVariable String serverName,
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:2002]: should be updateCategoryHierarchyLink

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java (hunk 3 above)
Comment (immmus) [reply open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:1438]: No, There is "got" here. Remove after on 1439 line.

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java (hunk 2 above)
Comment (immmus) [reply open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:205]: FVTUtils.checkGuidEnd1s("CategoryHierarchyLink", parent.getSystemAttributes().getGUID(), categoryHierarchyLink.getSuperCategoryGuid());
        FVTUtils.checkGuidEnd2s("CategoryHierarchyLink", child.getSystemAttributes().getGUID(), categoryHierarchyLink.getSubCategoryGuid());


Its check on 1456 and 1457 lines

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 7]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+     */
+    @PostMapping(path = "/users/{userId}/relationships/category-hierarchy-link")
+    public SubjectAreaOMASAPIResponse<CategoryHierarchyLink> createCategoryHierarchyLink(@PathVariable String serverName,
+                                                                                         @PathVariable String userId,
+                                                                                         @RequestBody CategoryHierarchyLink categoryHierarchyLink) {
+        return restAPI.createCategoryHierarchyLink(serverName, userId, categoryHierarchyLink);
+    }
+
+    /**
+     * Get a categoryAnchor Relationship. A relationship between two categories used to create nested categories.
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:1965]: categoryAnchor incorrect

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 8]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
-        return restoreLine(serverName, restAPIName, userId, SemanticAssignmentMapper.class, guid);
+        return getLine(serverName, restAPIName, userId, SemanticAssignmentMapper.class, guid);
+    }
+
+    /**
+     * Update a CategoryHierarchyLink Relationship. A relationship between a Category and a Category.
+     *
+     * @param serverName           serverName under which this request is performed, this is used in multi tenanting to identify the tenant
+     * @param userId               userId under which the request is performed
+     * @param guid       guid of the CategoryAnchor relationship
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1946]: CategoryAnchor incorrect and formatting

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 9]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
+     * @param categoryHierarchyLink the categoryAnchor relationship
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1947]: CategoryAnchor incorrect and formatting

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 10]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
+     * <li> PropertyServerException              Property server exception. </li>
+     * </ul>
+     */
+    public SubjectAreaOMASAPIResponse<CategoryHierarchyLink> createCategoryHierarchyLink(String serverName, String userId, CategoryHierarchyLink categoryHierarchyLink) {
+        String restAPIName = "createCategoryHierarchyLink";
+        return createLine(serverName,restAPIName,userId, CategoryHierarchyLinkMapper.class, categoryHierarchyLink);
+    }
+
+    /**
+     * Get a categoryAnchor Relationship. A relationship between a Category and a Category.
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1982]: CategoryAnchor incorrect

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 11]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
+     *
+     * @param serverName serverName under which this request is performed, this is used in multi tenanting to identify the tenant
+     * @param userId     unique identifier for requesting user, under which the request is performed
+     * @param guid       guid of the categoryAnchor Relationship to get
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1986]: CategoryAnchor incorrect

File: open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java [hunk 12]
```diff
@@ -1935,6 +1935,107 @@ public SubjectAreaRelationshipRESTServices() {
...
+     * @return response which when successful contains the categoryAnchorRelationship relationship with the requested guid
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-server/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/services/SubjectAreaRelationshipRESTServices.java:1987]: CategoryAnchor incorrect - I suggest we do not need the word relationship here

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 13]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+     *
+     * @param serverName serverName under which this request is performed, this is used in multi tenanting to identify the tenant
+     * @param userId     unique identifier for requesting user, under which the request is performed
+     * @param guid       guid of the categoryAnchor Relationship to get
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:1969]: CategoryAnchor incorrect

File: open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java [hunk 14]
```diff
@@ -1939,4 +1939,120 @@ public SubjectAreaRelationshipRESTResource() {
...
+     * @return response which when successful contains the categoryAnchorRelationship relationship with the requested guid
```
Comment (davidradl) [inline open-metadata-implementation/access-services/subject-area/subject-area-spring/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/server/spring/SubjectAreaRelationshipRESTResource.java:1970]: CategoryAnchor incorrect and formatting

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java (hunk 3 above)
Comment (davidradl) [reply open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:1445]: I am not sure I understand - I suggest we do not add commented out lines of code.

File: open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java [hunk 15]
```diff
@@ -1423,4 +1432,52 @@ private ProjectScope createProjectScope(Project project, Term term) throws Inval
...
+        Category gotChild = subjectAreaCategory.getByGUID(userId, child.getSystemAttributes().getGUID());
+        checkParent(parent, gotChild);
+        subjectAreaRelationship.categoryHierarchyLink().delete(this.userId, guid);
+        //FVTUtils.validateLine(gotTermCategorizationRelationship);
+        System.out.println("Soft deleted CategoryHierarchyLink with userId=" + guid);
+        gotCategoryHierarchyLink = subjectAreaRelationship.categoryHierarchyLink().restore(this.userId, guid);
+        FVTUtils.validateLine(gotCategoryHierarchyLink);
+        System.out.println("Restored CategoryHierarchyLink with userId=" + guid);
+        subjectAreaRelationship.categoryHierarchyLink().delete(this.userId, guid);
+        //FVTUtils.validateLine(gotTermCategorizationRelationship);
```
Comment (davidradl) [inline open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:1451]: I suggest we remove the commented out line of code we are adding

Comment (immmus) [reply open-metadata-test/open-metadata-fvt/access-services-fvt/subject-area-fvt/src/main/java/org/odpi/openmetadata/accessservices/subjectarea/fvt/RelationshipsFVT.java:1451]: I agree to you

**Comment by immmus** [comment]
Hi @davidradl 

When are you merge this?;)

**Pull request merged**